proxy.go

```diff
@@ -193,6 +242,229 @@ func (proxy *proxy) releaseToken(token Token) (*tokenInfo, error) {
+		},
+		Tokens: tokens,
+	}
+	o, err := json.MarshalIndent(&odVM, "", "\t")
+	if err != nil {
+		proxyLog.WithField("vm", vm.containerID).Warnf(
+			"Couldn't marshal VM state: %v", err)
+		return
+	}
+	storeFile := vmStateFilePath(vm.containerID)
+	err = ioutil.WriteFile(storeFile, o, proxyStateFilesPerm)
+	if err != nil {
+		proxyLog.WithField("vm", vm.containerID).Warnf(
+			"Couldn't store VM state to %s: %v", storeFile, err)
+	}
+}
+
+func delVMAndState(proxy *proxy, vm *vm) {
+	if proxy == nil {
+		proxyLog.Error("proxy parameter must be not nil")
+		return
+	}
+	if vm == nil {
+		proxyLog.Error("vm parameter must be not nil")
+		return
+	}
+	proxyLog.Infof("Removing on-disk state of %s", vm.containerID)
+	proxy.Lock()
+	delete(proxy.vms, vm.containerID)
+	proxy.Unlock()
+	proxy.storeState()
+	storeFile := vmStateFilePath(vm.containerID)
+	if err := os.Remove(storeFile); err != nil {
+		proxyLog.WithField("vm", vm.containerID).Warnf(
+			"Couldn't remove file %s: %v", storeFile, err)
+	}
+}
+
+func readVMState(containerID string) *vmStateOnDisk {
+	if containerID == "" {
+		proxyLog.Errorf("containerID parameter must be not empty")
+		return nil
+	}
+	vmStateFilePath := vmStateFilePath(containerID)
+	fdata, err := ioutil.ReadFile(vmStateFilePath)
+	if err != nil {
+		proxyLog.Errorf("Couldn't read %s: %v", vmStateFilePath, err)
+		return nil
+	}
+
+	var vmState vmStateOnDisk
+	err = json.Unmarshal(fdata, &vmState)
+	if err != nil {
+		proxyLog.Errorf("Couldn't unmarshal %s: %v", vmStateFilePath, err)
+		return nil
+	}
+	proxyLog.Debugf("restoring vm state: %+v", vmState)
+	return &vmState
+}
+
+func restoreTokens(proxy *proxy, vmState *vmStateOnDisk, vm *vm) {
+	if err := proxy.restoreTokens(vm, vmState.Tokens); err != nil {
+		proxyLog.Errorf("Failed to restore tokens: %v", err)
+		return
+	}
+
+	for _, token := range vmState.Tokens {
+		session := vm.findSessionByToken(Token(token))
+		if session == nil {
+			proxyLog.Errorf("Session must be not nil")
+			delVMAndState(proxy, vm)
+			return
+		}
+		if err := session.WaitForShim(); err != nil {
+			proxyLog.Errorf("Failed to re-connect with shim: %v", err)
+			delVMAndState(proxy, vm)
+			return
+		}
+	}
+}
+
+func restoreVMState(proxy *proxy, containerID string) {
+	if proxy == nil {
+		proxyLog.Errorf("proxy parameter must be not nil")
+		return
+	}
+
+	vmState := readVMState(containerID)
+	if vmState == nil {
+		return
+	}
+
+	regVM := vmState.RegisterVM
+	if regVM.ContainerID == "" || regVM.CtlSerial == "" || regVM.IoSerial == "" {
+		proxyLog.Errorf("wrong VM parameters")
+		return
+	}
+
+	proxy.Lock()
+	if _, ok := proxy.vms[regVM.ContainerID]; ok {
+		proxy.Unlock()
+		proxyLog.Errorf("%s: container already registered", regVM.ContainerID)
+		return
+	}
+	vm := newVM(regVM.ContainerID, regVM.CtlSerial, regVM.IoSerial)
+	proxy.vms[regVM.ContainerID] = vm
+	proxy.Unlock()
+
+	proxyLog.Infof("restoreVMState(containerId=%s,ctlSerial=%s,ioSerial=%s,console=%s)",
```

Comment:
It would be great if you could convert all the new log messages on this PR to use the structured logging approach. For example, this one could become:

```
proxyLog.WithFields(logrus.Fields{
  "container": regVM.ContainerID,
  "control-channel": regVM.CtlSerial,
  "io-channel": regVM.IoSerial,
  "console": regVM.Console,
}).Info("restoring state")
```

Reply:
Ok, will do.